tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,293 @@
+    \* NOTE: I think this is redundant and is not needed for this model.
+    dishonestSubnets |-> {}
+  ]
+
+Init ==
+  /\ LedgerInit
+  /\ SubnetsInit
+
+-----------------------------------------------------------------------------
+(***************************************************************************)
+(* Type checks                                                             *)
+(***************************************************************************)
+
+TransferMessage ==
+  \* A message to transfer cycles between subnets.
+  [type : {"transfer"}, from : SUBNETS, to : SUBNETS, amount : Nat \ {0}]
+
+\* The set of all possible messages that are sent/received by subnets.
+SubnetMessages ==
+  \* Transferring cycles from one subnet to another.
+  TransferMessage
+    \union
+    
+  \* Approving a cycles transfer.
+  \* This message is sent by the ledger to indicate that the `amount` sent is
+  \* now finalized.
+  [type : {"approve"}, from : SUBNETS, to : SUBNETS, amount : Nat \ {0}]  
+
+LedgerMessages ==
+  \* The set of all possible messages that can be received by the Ledger.
+  TransferMessage
+
+SubnetsTypeOK ==
+  /\ \A i \in DOMAIN subnetMsgs: {subnetMsgs[i]} \subseteq SubnetMessages
+  /\ subnets \in [SUBNETS -> [
+        balance: Nat,
+        unfinalized: [SUBNETS -> Nat],
+        msgsToLedger: Seq(LedgerMessages),
+        honest: {TRUE, FALSE}
+     ]]
+
+LedgerTypesOK ==
+  \* Type-correctness invariant.
+  /\ DOMAIN ledger.balances = SUBNETS
+  /\ \A s \in SUBNETS: ledger.balances[s] >= 0
+  /\ ledger.dishonestSubnets \subseteq SUBNETS
+  /\ \A i \in DOMAIN ledger.msgs: {ledger.msgs[i]} \subseteq LedgerMessages
+  
+TypesOK ==
+  \* Type-correctness invariant.
+  /\ LedgerTypesOK
+  /\ SubnetsTypeOK
+
+-----------------------------------------------------------------------------
+(***************************************************************************)
+(* Subnet actions                                                          *)
+(***************************************************************************)
+
+SubnetSendTransfer ==
+  (* Honestly sends a transfers cycles from one subnet to another. *)
+  
+  \* Choose two subnets.
+  \E sender, receiver \in SUBNETS: sender /= receiver
+  /\ subnets[sender].balance > 0
+  
+  \* Choose an amount that's within the subnet's balance.
+  /\ \E amount \in 1..subnets[sender].balance:
+  
+  \* Send a transfer to the receiver.
+  /\ subnetMsgs' = Append(subnetMsgs, [
+        type |-> "transfer",
+        from |-> sender,
+        to |-> receiver,
+        amount |-> amount
+     ])
+     
+  \* Subtract amount from sender's balance.
+  /\ subnets' = [subnets EXCEPT ![sender].balance = @ - amount]
+  
+  /\ UNCHANGED<<ledger>>
+
+SubnetDishonestSendTransfer ==
+  (* Maliciously transfers more cycles than the subnet's own balance. *)
+  
+  \E sender, receiver \in SUBNETS: sender /= receiver
+
+  \* To keep the state space bounded, a subnet can make a dishonest
```

Comment:
It may make more sense to limit the total amount of dishonest actions taken, instead of limiting the subnets to one dishonest action each; right now you would miss any attacks that would require the same subnet to do two dishonest things. I.e., have a counter of `total_dishonest_actions`, check that it's under some `MAX_DISHONEST_ACTIONS` and bump the counter here.